### ISSUE TITLE
Domainmodel Example: New line is inserted when executing 'Uncapitalize name' Quickfix

### ISSUE THREAD

**Issue by miklossy** [eclipse/xtext-eclipse#1221]
Given is the following `dmodel` file:

```
entity Blog {
	Title : String
}
```

When executing the Uncapitalize name on the property Title, an additional new line is inserted after the property:

![screencast](https://user-images.githubusercontent.com/20393472/66891455-80de9500-efe9-11e9-894e-53f8fef035de.gif)

See also the already defiend Domainmodel [QuickfixTest](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel.ui.tests/src/org/eclipse/xtext/example/domainmodel/ui/tests/QuickfixTest.xtend#L56).

**Comment by nbhusare** [comment]
I problem is caused by the code on line 58 of the DomainmodelFormatter class (https://bit.ly/2UgfjrM). Setting the value of the "defaultNewLines" param to 0 resolves the issue.  

I am not sure if the above change would have any side-effect.  @cdietrich Could you please confirm.

**Comment by miklossy** [comment]
@nbhusare ! Thanks for analysing this problem. Could you please verify if the additional newline is also inserted wenn invoking only the formatter without the quickfix?

**Comment by cdietrich** [comment]
i did some tests. they look good.
but i am not sure if my examples were sufficient.

i also dont know if this just fixes the symptoms and not the cause

**Comment by nbhusare** [comment]
@miklossy After making the change, I verified by manually invoking the formatter, and it looks good to me. I also had a feeling that the resolution only fixes the symptom, and hence I asked @cdietrich to confirm. Let me take some time to look into this properly.

**Comment by miklossy** [comment]
I analyzed the problem and found that the additional newline is inserted while serializing the modified property.  This happens in the [serializeReplacement](https://github.com/eclipse/xtext-core/blob/f9b6d1bebe09dc5775d45671d255384a8160015c/org.eclipse.xtext/src/org/eclipse/xtext/serializer/impl/Serializer.java#L207-L213) method. The `text` variable contains the `title : String` text, and additionally, due to the [DomainmodelFormatter defaultNewLines is set to 1](https://github.com/eclipse/xtext-eclipse/blob/master/org.eclipse.xtext.xtext.ui.examples/projects/domainmodel/org.eclipse.xtext.example.domainmodel/src/org/eclipse/xtext/example/domainmodel/formatting2/DomainmodelFormatter.xtend#L58), the serializer appends a new line after the text, resulting in `title : String\r\n`. The quickfix replaces the old text `Title : String` (calculated by `ICompositeNode`) by the new text `title : String\r\n`.

I propose the fix to modify the `ReplaceRegion length` calculation so that it takes not only the `ICompositeNode.getTotalLength()` into account, but considers if the new text contains additional whitespaces and the old text is also followed by white spaces, than the whitespaces contained by the original document is also consumed by the quickfix.

**Comment by miklossy** [comment]
I can also reproduce this problem based on the [NoJdtTestLanguage](https://github.com/eclipse/xtext-core/blob/master/org.eclipse.xtext.testlanguages/src/org/eclipse/xtext/testlanguages/noJdt/NoJdtTestLanguage.xtext) in the xtext-core repository, if extend its Formatter as follows:
```xtend
class NoJdtTestLanguageFormatter extends AbstractFormatter2 {
	
	...
	
	def dispatch void format(Greeting greeting, extension IFormattableDocument document) {
		greeting.append[setNewLines(1, 1, 2)]
	}
}
```

**Comment by miklossy** [comment]
Fixed in 2.22
